xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -386,4 +388,60 @@ void reportClientStats(String clusterName, String serverUri, LoadStatsStore load
    */
   void cancelClientStatsReport(String clusterName) {
   }
+
+  /**
+   * An {@link ObjectPool} holding reference and ref-count of an {@link XdsClient} instance.
+   * Once the ref-count gets back to zero, the xds client will be shutdown and later calls of {@lint
+   * #getObject()} will return null.
+   */
+  static final class RefCountedXdsClientObjectPool implements ObjectPool<XdsClient> {
+    private final Object lock = new Object();
+
+    @Nullable
+    private XdsClient xdsClient;
+    private int refCount;
+
+    RefCountedXdsClientObjectPool(XdsClient xdsClient) {
```

Comment:
Seems you could just require it to be Closeable, which would also make it easier to test (no need for a mock of a complex class).

Reply:
I don't quite understand your suggestion. What API signature you are suggesting?

Reply:
Would that be

```
class RefCountedXdsClientObjectPool<T extends Closeable> implements ObjectPool<T> {
  RefCountedXdsClientObjectPool(T t)
}
class XdsClient implements Closeable {}
```

?

Reply:
This seems moot now, since I was suggesting to construct the XdsClient directly in the pool. But yes, that was what I was saying was an option.